[flang][debug] Support ClassType.

abidh opened the pull request:
This PR adds the handling of `ClassType`. It is treated as pointer to the underlying type. Note that `ClassType` when passed to the function have double indirection so it is represented as pointer to type (compared to other types which may have a single indirection).

If `ClassType` wraps a pointer or allocatable then we take care to generate it as PTR -> type (and not PTR -> PTR -> type).

This is how it looks like in the debugger.

```
subroutine test_proc (this)
    class(test_type), intent (inout) :: this
    allocate (this%b (3, 2))
    call fill_array_2d (this%b)
    print *, this%a
end
```

```
(gdb) p this
$6 = (PTR TO -> ( Type test_type )) 0x2052a0
(gdb) p this%a
$7 = 0
(gdb) p this%b
$8 = ((1, 2, 3) (4, 5, 6))

```

abidh committed:
[flang][debug] Support ClassType.

This PR adds the handling of ClassType. It is treated as pointer to
the underlying type. Note that ClassType when passed to the function
have double indirection so it is represented as pointer to type
(and not type as a RecordType will be for example).

If ClassType wraps a pointer or allocatable then we take care to
avoid double indirection.

This is how it looks like in the debugger.

subroutine test_proc (this)
    class(test_type), intent (inout) :: this
    allocate (this%b (3, 2))
    call fill_array_2d (this%b)
    print *, this%a
end

(gdb) p this
$6 = (PTR TO -> ( Type test_type )) 0x2052a0
(gdb) p this%a
$7 = 0
(gdb) p this%b
$8 = ((1, 2, 3) (4, 5, 6))
(gdb)

jeanPerier commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -617,6 +617,20 @@ DebugTypeGenerator::convertType(mlir::Type Ty, mlir::LLVM::DIFileAttr fileAttr,
     return convertRecordType(recTy, fileAttr, scope, declOp);
   } else if (auto tupleTy = mlir::dyn_cast_if_present<mlir::TupleType>(Ty)) {
     return convertTupleType(tupleTy, fileAttr, scope, declOp);
+  } else if (auto classTy = mlir::dyn_cast_if_present<fir::ClassType>(Ty)) {
```
ClassType and BoxType are implemented in the same way in the runtime, the difference is that the dynamic type inside a ClassType may mismatch its static type.

You can actually merge the handling of ClassType and BoxType by replacing `auto boxTy = mlir::dyn_cast_or_null<fir::BoxType>(Ty)` with `auto boxTy = mlir::dyn_cast_or_null<fir::BaseBoxType>(Ty)`.

You would then get the static type description of the entity in the DWARF (which seems to be what is being done here).

abidh replied:
Thanks for your comments. I updated the patch to handle it with the BoxType. I also made another small change to handle cases like `CLASS(*)`.

jeanPerier reviewed (commented):
The `fir.class` type is used for polymorphic entities (including unlimited polymorphic entities, `CLASS(*)`).

Does DWARF provide support for querying the dynamic type at runtime?

I saw dwarf has a DW_TAG_dynamic_type tag, but that seems related to POINTER/ALLOCATABLE length parametrized derived types rather than polymorphism.

abidh committed:
Handle review comments.

Merged handling of ClassType with BoxType. Used boxTy.unwrapInnerType()
to avoid special casing for pointer or allocatable. Also improved
convertPointerLikeType to better handle if underlying type in null or
none.

abidh commented:
> DW_TAG_dynamic_type



> he `fir.class` type is used for polymorphic entities (including unlimited polymorphic entities, `CLASS(*)`).
> 
> Does DWARF provide support for querying the dynamic type at runtime?
> 
> I saw dwarf has a DW_TAG_dynamic_type tag, but that seems related to POINTER/ALLOCATABLE length parametrized derived types rather than polymorphism.

Sorry I missed this comment earlier. I agree that `DW_TAG_dynamic_type` does not seem to be intended for polymorphism. I have noticed that gfortran uses an artificial type in such cases which have a data, vptr and len field. I was thinking that we can also use a type that exposes `elem_len` and `type` field from the descriptor.

```
(gdb) ptype p_t2
type = Type __class__STAR_p
    PTR TO -> ( void :: _data )
    PTR TO -> ( Type __vtype__STAR :: _vptr )
    integer(kind=8) :: _len
End Type __class__STAR_p
```

jeanPerier commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -628,8 +633,13 @@ DebugTypeGenerator::convertType(mlir::Type Ty, mlir::LLVM::DIFileAttr fileAttr,
-  } else if (auto boxTy = mlir::dyn_cast_or_null<fir::BoxType>(Ty)) {
-    auto elTy = boxTy.getElementType();
+  } else if (auto boxTy = mlir::dyn_cast_or_null<fir::BaseBoxType>(Ty)) {
+    if (mlir::isa<fir::ClassType>(Ty))
+      return convertPointerLikeType(boxTy.unwrapInnerType(), fileAttr, scope,
+                                    declOp, /*genAllocated=*/false,
+                                    /*genAssociated=*/true);
```
I do not get why `ClassType` is bypassing the handling below made for BoxType and why `genAssociated` would be always true for them even when they are not POINTERs.

With the exception of the polymorphic aspect (the fact that the dynamic type may be an extension of the static type), which seems not straightforward to represent in DWARF, a fir.class is just like a fir.box. So the representation of the static type of a fir.box and fir.class should be the the same I think.

My comment about `CLASS(*)` may have mislead you here. I think it is OK to decide and agree how to represent those in a distinct patch, especially because they are also `TYPE(*)`, so `fir.box<none>` is also a thing.

abidh committed:
Handle review comments.

Dont special case ClassTyp but handle it like a BoxType. Also in default case, generate a pointer to underlying type instead of a PlaceHolder type.

jeanPerier reviewed (approved):
LGTM, thanks

tblah reviewed (approved):
LGTM
